### PR TITLE
Random mission generator

### PR DESCRIPTION
**Pull request by afsafzal**
Fixes the problems with random mission generator.

### CODE REVIEW

**Comment by afsafzal** [comment]
@ChrisTimperley This PR is asking for your attention 😄

File: houston/ardu/command_factory.py [hunk 1]
```diff
@@ -13,6 +14,24 @@
 logger.setLevel(logging.DEBUG)
 
 
+def circle_based_generator(cls, rng: random.Random) -> Command:
```
Comment (ChrisTimperley) [inline houston/ardu/command_factory.py]: Can you add a type annnotation for `cls`?

File: houston/ardu/command_factory.py [hunk 2]
```diff
@@ -13,6 +14,24 @@
...
+    (lat, lon) = (-35.3632607, 149.1652351)  # FIXME
+    heading = rng.uniform(0.0, 360.0)
+    dist = rng.uniform(0.0, 2.0)  # FIXME
+    params = {}
```
Comment (ChrisTimperley) [inline houston/ardu/command_factory.py]:
```
params = {p.name: p.generate(rng) for p in cls.parameters}
```

File: houston/ardu/command_factory.py [hunk 3]
```diff
@@ -27,6 +46,7 @@ def create_command(command: Dict[str, Any]) -> Type[Command]:
     except KeyError:
         msg = "missing 'id' field of Command"
         raise TypeError(msg)
+    generator = command.get('generator')
```
Comment (ChrisTimperley) [inline houston/ardu/command_factory.py]: What happens if a command doesn't have a generator? `generator` will be `None`, of course, but how will that affect the rest of the system?

File: houston/ardu/copter/goto.py [hunk 4]
```diff
@@ -61,6 +62,23 @@ class GoTo(Command):
         Idle
     ]
 
+    @classmethod
+    def generate(cls, rng: random.Random) -> 'GoTo':
```
Comment (ChrisTimperley) [inline houston/ardu/copter/goto.py]: This looks an awful lot like the code in `circle_based_generator`.

File: houston/command.py [hunk 5]
```diff
@@ -238,7 +238,10 @@ def uid(self) -> str:
         """
         The UID of the type of this command.
         """
-        return self.__class__.uid
+#        return self.__class__.uid
```
Comment (ChrisTimperley) [inline houston/command.py]: Let's keep this method how it was.

File: houston/command.py [hunk 6]
```diff
@@ -346,6 +349,14 @@ def to_message(self) -> Message:
     def __iter__(self) -> Iterator[Parameter]:
         yield from self.__class__.parameters
 
+    @classmethod
+    def generate(cls, rng: random.Random) -> 'Command':
+        params = {}
```
Comment (ChrisTimperley) [inline houston/command.py]: See `params` comment above.

File: houston/generator/base.py [hunk 7]
```diff
@@ -28,49 +34,42 @@ def __next__(self):
...
+        with self.__lock:
             g.tick()
             if g.exhausted():
                 raise StopIteration
             mission = self.__generator.generate_mission()
             g.tick()
             g.resource_usage.num_missions += 1
             mission_num = g.resource_usage.num_missions
-            print('Generated mission: {}'.format(mission_num))  # FIXME
+            logger.debug('Generated mission: {}'.format(mission_num))
```
Comment (ChrisTimperley) [inline houston/generator/base.py]:
```
logger.debug('Generated mission: %d', mission_num)
```

File: houston/generator/base.py [hunk 8]
```diff
@@ -175,12 +174,16 @@ def tick(self):
         self.__resource_usage.running_time = \
             timeit.default_timer() - self.__start_time
 
-    def execute_mission(self, mission, container):
+    def execute_mission(self,
+                        bz: BugZooClient,
```
[the remark this answers is not in the document]
Comment (afsafzal) [reply houston/generator/base.py]: The `MissionGenerator` does not necessarily need to run missions. Like for random generation, we don't need to execute. So I think passing the `BugZooClient` to  `execute_mission` makes more sense

[the remark this answers is not in the document]
Comment (ChrisTimperley) [reply houston/generator/base.py]: Good point.

**Pull request merged**